File: validation/reactor_mixing_timing.py
"""Time the complete pinned mixer on prepared models, checking every output."""
import argparse,ast,hashlib,importlib.util,json,statistics,time
from pathlib import Path
import cantera as ct
import cantera._cantera as compiled
import numpy as np

SOURCE_SHA256='87ccbfa629cb6cd494afae82d206c60b215261e16982675264db017be4314e11'
def sha(path):return hashlib.sha256(Path(path).read_bytes()).hexdigest()

def snapshot(ns):
    p=ns['mixer'].phase
    return dict(Y=p.Y.copy(),X=p.X.copy(),temperature=np.array([p.T]),pressure=np.array([p.P]),
        mass=np.array([ns['mixer'].mass]),density=np.array([p.density]),enthalpy=np.array([p.enthalpy_mass]),
        internal_energy=np.array([p.int_energy_mass]),entropy=np.array([p.entropy_mass]),gibbs=np.array([p.gibbs_mass]),
        cp=np.array([p.cp_mass]),cv=np.array([p.cv_mass]),mean_molecular_weight=np.array([p.mean_molecular_weight]),
        flows=np.array([ns[name].mass_flow_rate for name in ('mfc1','mfc2','outlet')]))

def run(args):
    args.output.mkdir(parents=True,exist_ok=False)
    source=args.cantera_source/'samples/python/reactors/mix1.py'
    if sha(source)!=SOURCE_SHA256 or ct.__version__!='4.0.0a2':
        raise ValueError('pinned Cantera4 mixer source/runtime required')
    ref=json.loads((args.reference/'reference.json').read_text())
    assert ref['passed'] and ref['provenance']['source']==sha(source)
    inputs=args.reference/'inputs'
    assert all(sha(inputs/name)==digest for name,digest in ref['provenance']['inputs'].items())
    environment_path=args.source_root/'validation/benchmark_environment.py'
    spec=importlib.util.spec_from_file_location('mixer_environment',environment_path)
    environment=importlib.util.module_from_spec(spec);spec.loader.exec_module(environment)
    threads_before=environment.verify_numerical_threads(set_accelerate=True)
    paths=[source,Path(__file__),environment_path,args.reference/'reference.json',args.reference/'original.npz']
    paths.extend(inputs/name for name in ref['provenance']['inputs'])
    inventory=lambda:{str(p):sha(p) for p in paths}
    libraries=lambda:{**environment.cantera_library_hashes(compiled.__file__),Path(compiled.__file__).name:sha(compiled.__file__)}
    before=inventory();library_before=libraries()
    assert library_before==ref['cantera_loaded_sha256']
    tree=ast.parse(source.read_text())
    solve=next(i for i,node in enumerate(tree.body) if isinstance(node,ast.Expr)
        and isinstance(node.value,ast.Call) and isinstance(node.value.func,ast.Attribute)
        and node.value.func.attr=='solve_steady')
    tree.body=tree.body[:solve+1]
    class Prepared(ast.NodeTransformer):
        replacements=0
        def visit_Call(self,node):
            if isinstance(node.func,ast.Attribute) and isinstance(node.func.value,ast.Name) and node.func.value.id=='ct' and node.func.attr=='Solution':
                self.replacements+=1
                assert len(node.args)==1 and node.args[0].value in ('air.yaml','gri30.yaml') and not node.keywords
                return ast.copy_location(ast.Subscript(value=ast.Name(id='prepared',ctx=ast.Load()),slice=node.args[0],ctx=ast.Load()),node)
            return self.generic_visit(node)
    transform=Prepared();tree=transform.visit(tree);ast.fix_missing_locations(tree)
    assert transform.replacements==2
    program=compile(tree,str(source),'exec')
    prepared={name:ct.Solution(str(inputs/name)) for name in ('air.yaml','gri30.yaml')}
    assert prepared['air.yaml'].n_species==8 and prepared['gri30.yaml'].n_species==53
    reference=np.load(args.reference/'original.npz',allow_pickle=False)
    def calculation():
        ns={'prepared':prepared};exec(program,ns);return ns
    times=[];saved={};report=dict(passed=False,source_before=before,libraries_before=library_before,
        threads_before=threads_before,cantera_version=ct.__version__,source_sha256=sha(source),
        host=environment.host_metadata(),scope='all original network construction and solve statements; two mechanism-file loads replaced by prepared models; printing/diagram excluded',
        repetitions=9)
    try:
        for i in range(10):
            start=time.perf_counter();ns=calculation();elapsed=time.perf_counter()-start
            if i:times.append(elapsed)
            values=snapshot(ns)
            saved.update({f'run_{i:02d}_{k}':v for k,v in values.items()})
            np.savez(args.output/'outputs.npz',**saved)
            for key,value in values.items():np.testing.assert_allclose(value,reference[key],rtol=2e-12,atol=1e-12)
        report.update(source_after=inventory(),libraries_after=libraries(),threads_after=environment.verify_numerical_threads())
        assert report['source_after']==before and report['libraries_after']==library_before
        report['passed']=True
    finally:
        report.update(warm_seconds=times,median_seconds=statistics.median(times) if times else None)
        (args.output/'timing.json').write_text(json.dumps(report,indent=2)+'\n')
    print(f'Complete Cantera mixer median: {report["median_seconds"]} s; all output replays pass')

if __name__=='__main__':
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--source-root',type=Path,default=Path(__file__).resolve().parents[1])
    for name in ('cantera-source','reference','output'):parser.add_argument('--'+name,type=Path,required=True)
    run(parser.parse_args())

File: validation/reactor_mixing_timing.jl
# Complete warm mixing-network calculation on prepared mechanisms.
using Arrhenius, LinearAlgebra, SHA, TOML, Libdl, Statistics
import Arrhenius.NPZ
length(ARGS)==4 || error("SOURCE_ROOT INPUT_DIR VALIDATED_NATIVE_DIR OUTPUT_DIR")
include(joinpath(ARGS[1],"example","reactors","mixing_solver.jl"))
include(joinpath(ARGS[1],"validation","numerical_threads.jl"))

function mixing_timing(source,input,reference,output)
    ispath(output) && error("refusing existing timing output")
    mkpath(output)
    filehash(path)=bytes2hex(sha256(read(path)))
    paths=[joinpath(input,name) for name in ("gri30.yaml","gri30.yaml.npz","air.yaml","air.yaml.npz")]
    append!(paths,[joinpath(source,"example","reactors",name) for name in ("mixing_solver.jl","network_cases.jl")])
    append!(paths,[@__FILE__,joinpath(source,"validation","numerical_threads.jl"),
        joinpath(reference,"native.npz"),joinpath(reference,"native.toml")])
    for (folder,_,files) in walkdir(joinpath(pkgdir(Arrhenius),"src")),name in files
        endswith(name,".jl") && push!(paths,joinpath(folder,name))
    end
    inventory()=Dict(path=>filehash(path) for path in paths)
    libraries()=Dict(basename(path)=>filehash(path) for path in Libdl.dllist()
        if isfile(path) && occursin(r"blas|mkl|accelerate|iomp|libgomp"i,basename(path)))
    before=inventory()
    threads_before=benchmark_julia_thread_settings(;enforce=true)
    gas=CreateSolution(joinpath(input,"gri30.yaml"));air=CreateSolution(joinpath(input,"air.yaml"))
    expected=NPZ.npzread(joinpath(reference,"native.npz"))
    meta=TOML.parsefile(joinpath(reference,"native.toml"))
    meta["converged"] && meta["inputs_unchanged"] || error("validated native calculation required")
    meta["native_sha256"]==filehash(joinpath(reference,"native.npz")) || error("native archive changed")
    meta["candidate_sha256"]==filehash(joinpath(source,"example","reactors","mixing_solver.jl")) || error("solver differs from validated calculation")
    for name in ("gri30.yaml","gri30.yaml.npz","air.yaml","air.yaml.npz")
        matches=[digest for (path,digest) in meta["inputs_before"] if basename(path)==name]
        length(matches)==1 && only(matches)==filehash(joinpath(input,name)) || error("mechanism differs from validated calculation")
    end
    function check(result)
        result.converged && result.physical_residual<=1e-9 || error("mixer failed convergence")
        result.iterations==meta["iterations"] || error("iteration count changed")
        isequal(result.state,expected["state"]) || error("stationary state changed")
        isequal(network_state(result.network),expected["initial_state"]) || error("initial state changed")
        diagnostic=network_diagnostics(result.network,result.state)
        isequal(diagnostic.mass_flow_rates,expected["flows"]) || error("boundary flow changed")
        isequal(diagnostic.nodes.mixer.mass_fractions,expected["Y"]) || error("species changed")
        nothing
    end
    # Specialize and warm the same complete call before timing.
    check(solve_mixing_network(gas,air))
    libraries_before=libraries()
    arrays=Dict{String,Any}();times=Float64[];iterations=Int[];residuals=Float64[]
    report=Dict{String,Any}("passed"=>false,"source_before"=>before,"threads_before"=>threads_before,
        "libraries_before"=>libraries_before,"scope"=>"fresh network construction and full steady solve; prepared mechanism loading and validation excluded",
        "julia_version"=>string(VERSION),"system"=>string(Sys.KERNEL),"cpu"=>Sys.cpu_info()[1].model,
        "kernel_release"=>Sys.iswindows() ? string(Sys.windows_version()) : readchomp(`uname -r`))
    try
        for i in 1:9
            start=time_ns()
            result=solve_mixing_network(gas,air)
            push!(times,(time_ns()-start)*1e-9)
            arrays["state_$i"]=copy(result.state)
            push!(iterations,result.iterations);push!(residuals,result.physical_residual)
            NPZ.npzwrite(joinpath(output,"states.npz"),arrays)
            check(result)
        end
        report["source_after"]=inventory()
        report["threads_after"]=benchmark_julia_thread_settings(;enforce=false)
        report["libraries_after"]=libraries()
        report["source_after"]==before || error("inputs changed during timing")
        report["libraries_after"]==libraries_before || error("loaded libraries changed during timing")
        any(path->occursin(r"cantera|coolprop|python"i,path),Libdl.dllist()) && error("foreign solver loaded")
        report["passed"]=true
    finally
        report["warm_seconds"]=times
        report["median_seconds"]=isempty(times) ? NaN : median(times)
        report["iterations"]=iterations;report["physical_residuals_per_s"]=residuals
        open(joinpath(output,"timing.toml"),"w") do io;TOML.print(io,report;sorted=true);end
    end
    println("Complete mixer median: ",median(times)," s; 9 exact output replays")
end
mixing_timing(ARGS...)
